test_locate_v1.test_blue_seen fails occasionally

The test ``test_blue_seen`` in ``ichnaea/api/locate/tests/test_locate_v1.py`` fails occasionally, such as a recent failure:

```
___________________________ TestView.test_blue_seen ____________________________
Traceback (most recent call last):
  File "/app/ichnaea/api/locate/tests/test_locate_v1.py", line 176, in test_blue_seen
    self.check_queue(data_queues, 0)
  File "/app/ichnaea/api/locate/tests/base.py", line 210, in check_queue
    assert data_queues["update_incoming"].size() == num
AssertionError
```

As [willkg says](https://github.com/mozilla/ichnaea/pull/920#issuecomment-534136454), it's weird that the assertion error shows like that and doesn't show the pytest fancy assertion stuff.

Here's the code:

https://github.com/mozilla/ichnaea/blob/7ee9de14a2b44456adf71b3d3113f5cbdf851bdf/ichnaea/api/locate/tests/test_locate_v1.py#L168-L176

The "v1" API is the current locate API, "v0" is the deprecated API. ``BlueShardFactory`` has some randomness for the MAC address, but the lat and long are constant:

https://github.com/mozilla/ichnaea/blob/5b020c4bd41246f3cefb98949a9555e598fe7e45/ichnaea/tests/factories.py#L143-L157

The frequency of failures may give some information, as well as if it is always this test that fails.

On Monday Oct 21, [build 1823](https://circleci.com/gh/mozilla/ichnaea/1823) with a traceback:

```
___________________________ TestView.test_blue_seen ____________________________
Traceback (most recent call last):
  File "/app/ichnaea/api/locate/tests/test_locate_v1.py", line 182, in test_blue_seen
    self.check_queue(data_queues, 0)
  File "/app/ichnaea/api/locate/tests/base.py", line 208, in check_queue
    assert data_queues["update_incoming"].size() == num
AssertionError: assert 1 == 0
 +  where 1 = <bound method DataQueue.size of <ichnaea.queue.DataQueue object at 0x7fb068d55890>>()
 +    where <bound method DataQueue.size of <ichnaea.queue.DataQueue object at 0x7fb068d55890>> = <ichnaea.queue.DataQueue object at 0x7fb068d55890>.size
```

So, we know the queue was empty at the start of the test, and there was exactly one item at the end of the test.

--- a/ichnaea/api/locate/cell.py
+++ b/ichnaea/api/locate/cell.py
@@ -1,5 +1,6 @@
 """Search implementation using a cell database."""
 
+from base64 import b64decode
 from collections import defaultdict
 import math
 
@@ -42,7 +43,7 @@ NETWORK_DTYPE = numpy.dtype(
         ("age", numpy.int32),
         ("signalStrength", numpy.int32),
         ("score", numpy.double),
-        ("id", "S11"),
+        ("id_b64", "S16"),
         ("seen_today", numpy.bool),
     ]
 )
@@ -79,7 +80,7 @@ def cluster_cells(cells, lookups, min_age=0):
                         obs_data[cell.cellid][0],
                         obs_data[cell.cellid][1],
                         station_score(cell, now),
-                        encode_cellid(*cell.cellid),
+                        encode_cellid(*cell.cellid, codec="base64"),
                         bool(cell.last_seen is not None and cell.last_seen >= today),
                     )
                     for cell in area_cells
@@ -118,7 +119,7 @@ def cluster_areas(areas, lookups, min_age=0):
                         obs_data[area.areaid][0],
                         obs_data[area.areaid][1],
                         area_score(area, now),
-                        encode_cellarea(*area.areaid),
+                        encode_cellarea(*area.areaid, codec="base64"),
                         bool(area.last_seen is not None and area.last_seen >= today),
                     )
                 ],
@@ -284,8 +285,8 @@ class CellPositionMixin(object):
                     )
 
                     used_networks = [
-                        ("cell", bytes(id_), bool(seen_today))
-                        for id_, seen_today in cluster[["id", "seen_today"]]
+                        ("cell", b64decode(id_b64), bool(seen_today))
+                        for id_b64, seen_today in cluster[["id_b64", "seen_today"]]
                     ]
 
                     results.add(
@@ -312,8 +313,8 @@ class CellPositionMixin(object):
                     )
 
                     used_networks = [
-                        ("area", bytes(id_), bool(seen_today))
-                        for id_, seen_today in cluster[["id", "seen_today"]]
+                        ("area", b64decode(id_b64), bool(seen_today))
+                        for id_b64, seen_today in cluster[["id_b64", "seen_today"]]
                     ]
 
                     results.add(

--- a/ichnaea/api/locate/internal.py
+++ b/ichnaea/api/locate/internal.py
@@ -56,10 +56,10 @@ class InternalPositionSource(
             return
 
         result_networks = {"area": set(), "blue": set(), "cell": set(), "wifi": set()}
-        for network in best_result.used_networks:
-            if network[2]:
+        for net_type, net_id, seen_today in best_result.used_networks:
+            if seen_today:
                 # only add network if it was last_seen today
-                result_networks[network[0]].add(network[1])
+                result_networks[net_type].add(net_id)
 
         if result_networks == query.networks():
             # don't store queries, based exclusively on data,

--- a/ichnaea/api/locate/mac.py
+++ b/ichnaea/api/locate/mac.py
@@ -1,5 +1,6 @@
 """Search implementation using a mac based source."""
 
+from base64 import b64decode
 from collections import defaultdict
 import itertools
 import math
@@ -22,7 +23,7 @@ NETWORK_DTYPE = numpy.dtype(
         ("age", numpy.int32),
         ("signalStrength", numpy.int32),
         ("score", numpy.double),
-        ("mac", "S6"),
+        ("mac_b64", "S8"),
         ("seen_today", numpy.bool),
     ]
 )
@@ -55,7 +56,7 @@ def cluster_networks(
                 obs_data[model.mac][0],
                 obs_data[model.mac][1],
                 station_score(model, now),
-                encode_mac(model.mac),
+                encode_mac(model.mac, codec="base64"),
                 bool(model.last_seen is not None and model.last_seen >= today),
             )
             for model in models
@@ -181,8 +182,8 @@ def aggregate_cluster_position(
     score = float(cluster["score"].sum())
 
     used_networks = [
-        (data_type, bytes(mac), bool(seen_today))
-        for mac, seen_today in sample[["mac", "seen_today"]]
+        (data_type, b64decode(mac_b64), bool(seen_today))
+        for mac_b64, seen_today in sample[["mac_b64", "seen_today"]]
     ]
 
     return result_type(

--- a/ichnaea/api/locate/tests/test_locate_v1.py
+++ b/ichnaea/api/locate/tests/test_locate_v1.py
@@ -170,13 +170,17 @@ class TestView(LocateV1Base, CommonLocateTest, CommonPositionTest):
             }
         ]
 
-    def test_blue_seen(self, app, data_queues, session):
+    @pytest.mark.parametrize("mac", ["a82067491500", "a82067491501"])
+    def test_blue_seen(self, app, data_queues, session, mac):
         """If a query contains no new data, it is not queued for further processing.
 
-        This fails occasionally, see issue #921.
+        Due to an issue with truncating bytestrings in a numpy array (numpy
+        issue 8089), addresses that end in '00' were once detected as new
+        stations.
         """
+
         self.check_queue(data_queues, 0)
-        blue = BlueShardFactory()
+        blue = BlueShardFactory(mac=mac)
         offset = 0.00002
         blues = [blue, BlueShardFactory(lat=blue.lat + offset)]
         session.flush()

